### PR TITLE
drop use of upstream zuul role, make our own

### PR DESCRIPTION
**Pull request by gandelman-a**

### CODE REVIEW

File: roles/zuul/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  notify: Restart zuul
+
+- name: Install layout.yaml
+  become: yes
+  template:
+    src: etc/zuul/config/layout.yaml.j2
+    dest: /etc/zuul/config/layout.yaml
+    owner: zuul
+  notify: Restart zuul
```
Comment (greghaynes) [inline roles/zuul/tasks/main.yml]: We dont need this - zuul will auto reload its config

File: roles/zuul/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+- name: Install service environment configs
+  become: yes
+  template:
+    src: etc/default/zuul
+    dest: /etc/default/zuul
+  notify: Restart zuul
```
Comment (greghaynes) [inline roles/zuul/tasks/main.yml]: I think we dont want to restart zuul anywhere in here - its going to be disruptive so were best off doing it separately IMO (when were ready for a downtime / do the save state dance).

File: roles/zuul/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    dest: "/etc/systemd/system/{{ item }}.service"
+  with_items:
+    - zuul-launcher
+    - zuul-merger
+    - zuul-server
+  notify: Restart zuul
+
+- name: Reload systemd
+  become: yes
+  command: systemctl daemon-reload
```
Comment (jamielennox) [inline roles/zuul/tasks/main.yml]: nit: want a 

`register: systemd_unit_files`

then 

`when systemd_unit_files.changed`

here. Alternatively there is a systemd module you can use instead of service that can be told to do this.

File: roles/zuul/tasks/main.yml (hunk 2 above)
Comment (jamielennox) [inline roles/zuul/tasks/main.yml]: This one probably is fine as it's going to only change when you are resetting the whole daemon - but whatever i see your point.

File: roles/zuul/handlers/main.yml [hunk 4]
```diff
@@ -1,10 +0,0 @@
-# We may not want to restart all services?
-- name: Restart zuul
-  become: yes
-  service:
-    name: "{{ item }}"
-    state: restarted
-  with_items:
-    - zuul-launcher
-    - zuul-merger
-    - zuul-server
```
Comment (gandelman-a) [inline roles/zuul/handlers/main.yml]: Im not entirely clear on what can and cannot be restarted without breaking state.  I'm willing to punt on this for now until we have things running and we can see exactly what the implications are.

File: roles/zuul/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,111 @@
+---
+- name: Install required packages
+  become: yes
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Can we just run the play with `become: yes` instead of tacking it onto every task?

File: roles/zuul/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+- name: Create directories
+  become: yes
+  file:
+    path: "{{ item }}"
+    state: directory
+    mode: 0755
+    owner: zuul
+  items:
+    - /etc/zuul
+    - /etc/zuul/config
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Ansible does leading directories automatically, so we don't have to put the base `/etc/zuul` here, it'll get created as part of `/etc/zuul/config`.

File: roles/zuul/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  with_items:
+    - /opt/source
+    - "{{ zuul_venv_dir }}"
+
+- name: Checkout zuul
+  become: yes
+  git:
+    dest: /opt/source/zuul
+    repo: "{{ zuul_git_repo_url }}"
+    version: "{{ zuul_git_branch }}"
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Should change in code cause new install/restart of zuul?

File: roles/zuul/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+- name: Install zuul
+  become: yes
+  pip:
+    name: file:///opt/source/zuul
+    virtualenv: "{{ zuul_venv_dir }}"
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Should this run every time, or only when the code changed in the previous task?

File: roles/zuul/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+- name: Install configs
+  become: yes
+  template:
+    src: "etc/zuul/{{ item }}.j2"
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Please drop the `.j2`

File: roles/zuul/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    dest: "/etc/zuul/{{ item }}"
+    owner: zuul
+  with_items:
+    - zuul.conf
+    - gearman-logging.conf
+    - launcher-logging.conf
+    - merger-logging.conf
+    - server-logging.conf
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Should we restart on code change (or if it's not safe yet, in the handler itself have a conditional that defaults to not restarting)?

File: roles/zuul/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+- name: Install layout.yaml
+  become: yes
+  template:
+    src: etc/zuul/config/layout.yaml.j2
+    dest: /etc/zuul/config/layout.yaml
+    owner: zuul
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Same, restarts, drop .j2

File: roles/zuul/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+- name: Install service environment configs
+  become: yes
+  template:
+    src: etc/default/zuul
+    dest: /etc/default/{{ item }}
+  with_items:
+    - zuul-launcher
+    - zuul-merger
+    - zuul-server
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: These seem like things that should trigger a restart too should they ever change.

File: roles/zuul/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    src: etc/systemd/system/zuul.service
+    dest: /etc/systemd/system/{{ item }}.service
+  with_items:
+    - zuul-launcher
+    - zuul-merger
+    - zuul-server
+
+- name: Reload systemd
+  become: yes
+  command: systemctl daemon-reload
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Can this be a handler like in the node pool role?

File: roles/zuul/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Handlers (once we add them) should be flushed before starting the service.

File: roles/zuul/templates/etc/zuul/config/layout.yaml.j2 [hunk 15]
```diff
@@ -0,0 +1,8 @@
+# This file is generated by Ansible
+#  DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
```
Comment (j2sol) [inline roles/zuul/templates/etc/zuul/config/layout.yaml.j2]: This should be `{{ ansible_managed }}`.

File: roles/zuul/templates/etc/zuul/gearman-logging.conf.j2 [hunk 16]
```diff
@@ -0,0 +1,39 @@
+[loggers]
```
Comment (j2sol) [inline roles/zuul/templates/etc/zuul/gearman-logging.conf.j2]: Is there significant difference in these files that can't be handled with templating of a single file?

File: roles/zuul/templates/etc/zuul/zuul.conf.j2 [hunk 17]
```diff
@@ -1,5 +1,3 @@
-# {{ ansible_managed }}
```
Comment (j2sol) [inline roles/zuul/templates/etc/zuul/zuul.conf.j2]: Why is this removed? Is this file not going to be ansible managed?

File: roles/zuul/tasks/main.yml (hunk 8 above)
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: Also the file:// part isn't needed.

File: roles/zuul/handlers/main.yml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+---
+- name: Restart zuul
+  service:
+    name: "{{ item }}"
+    state: restarted
+  with_items:
+    - zuul-launcher
+    - zuul-merger
+    - zuul-server
+  when: zuul_allow_restart_services
```
Comment (j2sol) [inline roles/zuul/handlers/main.yml]: Because this will be set potentially by a human, we need to run this through the bool filter.

`when: zuul_allow_restart_services | bool`

File: roles/zuul/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+  notify: Restart zuul
+
+- name: Install missing pip dependencies
+  pip:
+    name: "{{ item.name }}"
+    version: "{{ item.version|default(omit) }}"
+    virtualenv: "{{ zuul_venv_dir }}"
+  with_items:
+    - { name: 'jenkins-job-builder', version: '1.6.1' }
+    - { name: 'pyzmq' }
```
Comment (j2sol) [inline roles/zuul/tasks/main.yml]: This will probably show unnecessary change each run, and could be triggered on the git change as well, but not necessary.

File: roles/zuul/handlers/main.yml [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+- name: Reload systemd units
```
Comment (j2sol) [inline roles/zuul/handlers/main.yml]: Since we want the proper systemd units in place before restarting the service, I think this task needs to be listed above the other.

Comment (jamielennox) [inline roles/zuul/handlers/main.yml]: this is a handler, doesn't it only matter the order you notify the handlers in?
